src/nanoemoji/nanoemoji.py

```diff
@@ -24,497 +24,135 @@
-        )
-
-    if not ttfont:
-        raise ValueError(
-            f"Unable to generate {config.color_format} {config.output_format}"
-        )
-
-    # Permit fixups where we can't express something adequately in UFO
-    _COLOR_FORMAT_GENERATORS[config.color_format].apply_ttfont(
-        ufo, color_glyphs, ttfont
-    )
+# internal flags, typically client wouldn't change
+flags.DEFINE_string("build_dir", "build/", "Where build runs.")
+flags.DEFINE_bool("gen_ninja", True, "Whether to regenerate build.ninja")
+flags.DEFINE_bool("exec_ninja", True, "Whether to run ninja.")
 
-    return ttfont
 
+def self_dir() -> str:
+    return os.path.dirname(os.path.abspath(__file__))
 
-def _write(ufo, ttfont, output_file):
-    logging.info("Writing %s", output_file)
 
-    if os.path.splitext(output_file)[1] == ".ufo":
-        ufo.save(output_file, overwrite=True)
-    else:
-        ttfont.save(output_file)
+def build_dir() -> str:
+    return os.path.abspath(FLAGS.build_dir)
 
 
-def _not_impl(*_):
-    raise NotImplementedError("%s not implemented" % FLAGS.color_format)
+def rel_self(path: str) -> str:
+    path = os.path.normpath(os.path.join(self_dir(), path))
+    return os.path.relpath(path, self_dir())
 
 
-def _next_name(ufo: ufoLib2.Font, name_fn) -> str:
-    i = 0
-    while name_fn(i) in ufo:
-        i += 1
-    return name_fn(i)
+def rel_build(path: str) -> str:
+    return os.path.relpath(path, build_dir())
 
 
-def _create_glyph(color_glyph: ColorGlyph, painted_layer: PaintedLayer) -> Glyph:
-    ufo = color_glyph.ufo
+def resolve_rel_build(path):
+    return os.path.abspath(os.path.join(build_dir(), path))
 
-    glyph = ufo.newGlyph(_next_name(ufo, lambda i: f"{color_glyph.glyph_name}.{i}"))
-    glyph_names = [glyph.name]
-    glyph.width = ufo.info.unitsPerEm
 
-    svg_units_to_font_units = color_glyph.transform_for_font_space()
+def write_preamble(nw):
+    def module_rule(mod_name, arg_pattern):
+        nw.rule(mod_name, f"{sys.executable} -m nanoemoji.{mod_name} {arg_pattern}")
 
-    if painted_layer.reuses:
-        # Shape repeats, form a composite
-        base_glyph = ufo.newGlyph(
-            _next_name(ufo, lambda i: f"{glyph.name}.component.{i}")
-        )
-        glyph_names.append(base_glyph.name)
+    nw.comment("Generated by nanoemoji")
+    nw.newline()
 
-        draw_svg_path(painted_layer.path, base_glyph.getPen(), svg_units_to_font_units)
+    nw.rule("picosvg", 'picosvg $in > $out || echo "$in failed picosvg"')
+    module_rule("write_codepoints", "$in > $out")
+    module_rule("write_fea", "$in > $out")
 
-        glyph.components.append(
-            Component(baseGlyph=base_glyph.name, transformation=Affine2D.identity())
-        )
+    keep_glyph_names = " --"
+    if not FLAGS.keep_glyph_names:
+        keep_glyph_names += "no"
+    keep_glyph_names += "keep_glyph_names"
+    module_rule(
+        "write_font",
+        " --codepoint_csv codepointmap.csv"
+        + " --fea_file features.fea"
```

Comment:
if these "codepointmap.csv" and "features.fea" are hard-coded, why do I see them among other flags when I do `nanoemoji --help`?
Are they meant to be overridden by user, if so then you should use the names provided by the caller of nanoemoji CLI.

Also, I would expect that if a codepointmap or features.fea are provided via explicit flags, they are not generated automatically but simply used as they are.

Reply:
I like the flags but it's making ninja slightly awkward because it just wants to list them as inputs. I'm going to move write_font to use positional / name pattern based args, remove the flags, and list the codepointmap and feature file as explicit inputs. That should fix the build graph and the messy --help.

Reply:
Build file looks much cleaner, ty, good feedback.